### PR TITLE
No more schema copying

### PR DESCRIPTION
**Pull request by paulbalaji**
#### Description

- no more copying around of schema
- updated feature module schema paths / namespaces
- contains a fix for https://github.com/spatialos/gdk-for-unity/issues/955
- should be friendly with future npm changes
- cleaned up some remaining mobile prefabs
- pinned shared-ci version

#### Tests

- repro steps for UTY-2044 no longer crash Unity

#### Documentation

- [x] will check if docs need to be updated
- [x] changelog

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.tools/Common.cs [hunk 1]
```diff
@@ -15,10 +13,10 @@ namespace Improbable.Gdk.Tools
     /// </summary>
     public static class Common
     {
-        /// <summary>
+        /// <summary>
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Common.cs]: Nit: needs a format pass

File: workers/unity/Packages/com.improbable.gdk.tools/Common.cs [hunk 2]
```diff
@@ -73,14 +70,25 @@ public static string GetPackagePath(string packageName)
...
+            {
+                return package.resolvedPath;
+            }
 
-            if (package == null)
+            // Package indirectly referenced (dependency)
+            var cachedPackage = Directory
+                .GetDirectories("Library/PackageCache")
+                .FirstOrDefault(path => Path.GetFileName(path).StartsWith($"{packageName}@"));
+            if (cachedPackage != null)
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Common.cs]: Prefer `!string.IsNullOrEmpty`

File: workers/unity/Packages/com.improbable.gdk.tools/GenerateCode.cs [hunk 3]
```diff
@@ -195,11 +190,24 @@ private static string[] ConstructArgs(string projectPath, string schemaCompilerP
             baseArgs.Add($"--native-output-dir=\"{toolsConfig.CodegenOutputDir}\"");
             baseArgs.Add($"--schema-path=\"{toolsConfig.SchemaStdLibDir}\"");
 
+            // Add user defined schema directories
             foreach (var schemaSourceDir in toolsConfig.SchemaSourceDirs)
             {
-                baseArgs.Add($"--schema-path=\"{schemaSourceDir}\"");
+                if (Directory.Exists(schemaSourceDir))
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/GenerateCode.cs]: I wonder if it's better to show an error in the configuration UI rather than silently ignoring invalid inputs here.

File: workers/unity/Packages/com.improbable.gdk.tools/GenerateCode.cs [hunk 4]
```diff
@@ -195,11 +190,24 @@ private static string[] ConstructArgs(string projectPath, string schemaCompilerP
...
+                {
+                    baseArgs.Add($"--schema-path=\"{schemaSourceDir}\"");
+                }
+            }
+
+            // Add package schema directories
+            foreach (var directory in GetSchemaDirectories())
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/GenerateCode.cs]: Could rewrite here (and above) as something like 
`baseArgs.AddRange(GetSchemaDirectories().Select(dir => $"--schema-path=\"{directory}\""));`

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 5]
```diff
@@ -204,7 +204,7 @@ public static void LaunchLocalDeployment()
             BuildConfig();
 
             var command = Common.SpatialBinary;
-            var commandArgs = "local launch";
+            var commandArgs = "local launch --enable_pre_run_check=false";
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Is this definitely needed?

Comment (paulbalaji) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: yes because it will otherwise call `prepare-for-run` under the hood - which will not work if schema is not in the default place (e.g. within the packages)

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,13 +2,20 @@
 
 ## Unreleased
 
+### Breaking Changes
+
+- Schema from packages are no longer copied into the root `schema` directory.
```
Comment (jamiebrynes7) [inline CHANGELOG.md]: Explain how to migrate their user schema if they previous did:

```
import from_gdk_packages/... 
```

File: ci/bootstrap.sh [hunk 7]
```diff
@@ -1,15 +1,22 @@
 #!/usr/bin/env bash
 set -e -u -x -o pipefail
 
-cd "$(dirname "$0")/../"
```
Comment (jamiebrynes7) [inline ci/bootstrap.sh]: Why is this removed?

File: test-project/Assets/Config/GdkToolsConfiguration.json [hunk 8]
```diff
@@ -1,7 +1,11 @@
 {
-    "SchemaStdLibDir": "build/dependencies/schema/standard_library",
+    "SchemaStdLibDir": "../../build/dependencies/schema/standard_library",
```
Comment (jamiebrynes7) [inline test-project/Assets/Config/GdkToolsConfiguration.json]: This puts it outside of your project directory I think: 

```
gdk-for-unity/
	test-project/
		build/ <-- Previous location
build/ <-- New location
```

File: workers/unity/Assets/Playground/Resources/Prefabs/Common/MobileConnectionSplashScreen.prefab [hunk 9]
```diff
@@ -1,970 +0,0 @@
-%YAML 1.1
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Playground/Resources/Prefabs/Common/MobileConnectionSplashScreen.prefab]: Uhhh intentional?

File: workers/unity/Packages/com.improbable.gdk.core/package.json [hunk 10]
```diff
@@ -1,13 +1,13 @@
...
-  "version": "0.2.2",
-  "unity": "2018.3",
-  "author": "Improbable Worlds Ltd",
-  "description": "SpatialOS GDK Core Module. Made by Improbable.",
-  "dependencies": {
-    "com.improbable.gdk.tools": "0.2.2",
-    "com.improbable.gdk.testutils": "0.2.2",
-    "com.unity.entities": "0.0.12-preview.21"
-  }
+    "name": "com.improbable.gdk.core",
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/package.json]: Our `.editorconfig` states that the indents are 2 spaces

File: workers/unity/Packages/com.improbable.gdk.core/package.json [hunk 11]
```diff
@@ -1,13 +1,13 @@
...
+    "displayName": "SpatialOS GDK Core",
+    "version": "0.2.2",
+    "unity": "2018.3",
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/package.json]: Driveby - these should be changed to `2019.1` 😉

File: workers/unity/Packages/com.improbable.gdk.tools/Common.cs [hunk 12]
```diff
@@ -36,33 +34,32 @@ public static readonly string
...
-                Debug.LogException(e);
-            }
-        }
-
         internal static string GetThisPackagePath()
         {
             return GetPackagePath("com.improbable.gdk.tools");
         }
 
+        static Common()
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/Common.cs]: Any reason you moved this around?

File: test-project/Assets/Config/GdkToolsConfiguration.json (hunk 8 above)
Comment (paulbalaji) [reply test-project/Assets/Config/GdkToolsConfiguration.json]: you are correct

made a mistake when copying across from playground

File: workers/unity/Packages/com.improbable.gdk.core/package.json (hunk 10 above)
Comment (paulbalaji) [reply workers/unity/Packages/com.improbable.gdk.core/package.json]: yeah I'm not sure what happened here - will drive by all the package.json files

File: workers/unity/Packages/com.improbable.gdk.tools/Common.cs (hunk 12 above)
Comment (paulbalaji) [reply workers/unity/Packages/com.improbable.gdk.tools/Common.cs]: Rider did it when I formatted the file

File: workers/unity/Assets/Playground/Resources/Prefabs/Common/MobileConnectionSplashScreen.prefab (hunk 9 above)
Comment (paulbalaji) [reply workers/unity/Assets/Playground/Resources/Prefabs/Common/MobileConnectionSplashScreen.prefab]: yes

Jess removed the screen in a different PR but forgot to remove the related prefab

File: workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfiguration.cs [hunk 13]
```diff
@@ -47,7 +50,29 @@ internal List<string> Validate()
...
+            {
+                errors.Add($"{GdkToolsConfigurationWindow.DescriptorOutputDirLabel} cannot be empty.");
+            }
+
+            var emptySchemaSourceDir = false;
+            foreach (var schemaSourceDir in SchemaSourceDirs)
+            {
+                if (!string.IsNullOrEmpty(schemaSourceDir))
+                {
+                    var fullSchemaSourceDirPath = Path.Combine(UnityProjectRoot, schemaSourceDir);
```
[the remark this answers is not in the document]
Comment (paulbalaji) [reply workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfiguration.cs]: looks like `Path.Combine` throws an `ArgumentException` if any invalid characters are used

so might be good to catch that + add a suitable error

File: workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfiguration.cs [hunk 14]
```diff
@@ -55,26 +54,32 @@ internal List<string> Validate()
...
+                    {
+                        fullSchemaSourceDirPath = Path.Combine(UnityProjectRoot, schemaSourceDir);
+                    }
+                    catch (ArgumentException)
+                    {
+                        errors.Add($"Schema path [{i}] contains one or more invalid characters.");
+                        continue;
+                    }
+
                     if (!Directory.Exists(fullSchemaSourceDirPath))
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfiguration.cs]: Move this into the try? Then you don't have to forward declare the `fullSchemaSourceDirPath` ?

File: workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfiguration.cs [hunk 15]
```diff
@@ -55,26 +54,32 @@ internal List<string> Validate()
                 errors.Add($"{GdkToolsConfigurationWindow.DescriptorOutputDirLabel} cannot be empty.");
             }
 
-            var emptySchemaSourceDir = false;
-            foreach (var schemaSourceDir in SchemaSourceDirs)
+            for (var i = 0; i < SchemaSourceDirs.Count; i++)
             {
+                var schemaSourceDir = SchemaSourceDirs[i];
+
                 if (!string.IsNullOrEmpty(schemaSourceDir))
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfiguration.cs]: Invert the `if` condition and continue if its null or empty: 

```csharp
if (string.IsNullOrEmpty(schemaSourceDir))
{
	errors.Add(...);
	continue;
}

try 
{
	...
}

...

```

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Components/BufferedTransform.cs [hunk 16]
```diff
@@ -7,7 +7,7 @@ public struct BufferedTransform : IBufferElementData
     {
         public Vector3 Position;
         public Vector3 Velocity;
-        public Quaternion Orientation;
+        public UnityEngine.Quaternion Orientation;
```
[the remark this answers is not in the document]
Comment (paulbalaji) [reply workers/unity/Packages/com.improbable.gdk.transformsynchronization/Components/BufferedTransform.cs]: yes, to avoid conflicts with the `Quaternion` defined in the transform module schema - which now gets generated in the `Improbable.Gdk.TransformSynchronization` namespace

File: workers/unity/Packages/com.improbable.gdk.tools/core-sdk.version [hunk 17]
```diff
@@ -1 +1 @@
-13.7.1-gdk-for-unity
+13.7.1-gdk-for-unity
```
[the remark this answers is not in the document]
Comment (paulbalaji) [reply workers/unity/Packages/com.improbable.gdk.tools/core-sdk.version]: not intended - but shouldn't make a difference

can make a mental note to remove new line for the next PR if needed
